pkg/tests/end_to_end_tests/tag_op_test.go

```diff
@@ -0,0 +1,460 @@
+    pg_typeof(x.op) as return_type
+from
+(
+	select ('tag1' %s %s) as op
+) x
+`
+	for _, opr := range operators {
+		for i, tval := range testValues {
+			testTagOpExpression(t, db, ctx, qry, opr, tval, i)
+		}
+	}
+}
+
+func testTagOpRegexpMatches(t *testing.T, db *pgxpool.Pool, ctx context.Context) {
+	const qry = `
+select 
+    (x.op).tag_key,
+    (x.op).value as output_value,
+    'text' as output_type,
+    pg_typeof(x.op) as return_type
+from
+(
+	select ('tag1' %s %s) as op
+) x
+`
+	opr := operator{
+		operator:   "==~",
+		returnType: "tag_op_regexp_matches",
+	}
+	tval := testValue{
+		inputType:   "text",
+		inputValue:  `'^\\d+$'`,
+		outputType:  "text",
+		outputValue: `^\\d+$`,
+	}
+	testTagOpExpression(t, db, ctx, qry, opr, tval, 1)
+}
+
+func testTagOpRegexpNotMatches(t *testing.T, db *pgxpool.Pool, ctx context.Context) {
+	const qry = `
+select 
+    (x.op).tag_key,
+    (x.op).value as output_value,
+    'text' as output_type,
+    pg_typeof(x.op) as return_type
+from
+(
+	select ('tag1' %s %s) as op
+) x
+`
+	opr := operator{
+		operator:   "!=~",
+		returnType: "tag_op_regexp_not_matches",
+	}
+	tval := testValue{
+		inputType:   "text",
+		inputValue:  `'^\\d+$'`,
+		outputType:  "text",
+		outputValue: `^\\d+$`,
+	}
+	testTagOpExpression(t, db, ctx, qry, opr, tval, 1)
+}
+
+func testTagOpJsonPathExists(t *testing.T, db *pgxpool.Pool, ctx context.Context) {
+	const qry = `
+select 
+    (x.op).tag_key,
+    (x.op).value as output_value,
+    'jsonpath' as output_type,
+    pg_typeof(x.op) as return_type
+from
+(
+	select ('tag1' %s %s) as op
+) x
+`
+	opr := operator{
+		operator:   "@?",
+		returnType: "tag_op_jsonb_path_exists",
+	}
+	tval := testValue{
+		inputType:   "jsonpath",
+		inputValue:  `'$ ? (@ > 3 && @ < 5)'::jsonpath`,
+		outputType:  "jsonpath",
+		outputValue: `$?(@ > 3 && @ < 5)`,
+	}
+	testTagOpExpression(t, db, ctx, qry, opr, tval, 1)
+}
+
+func testTagOpExpression(t *testing.T, db *pgxpool.Pool, ctx context.Context, qry string, opr operator, tval testValue, i int) {
+	name := fmt.Sprintf("%s_%s_%d", opr.returnType, tval.inputType, i)
+	t.Run(name, func(t *testing.T) {
+		row := db.QueryRow(ctx, fmt.Sprintf(qry, opr.operator, tval.inputValue))
+		var (
+			tagKey      string
+			outputValue string
+			outputType  string
+			returnType  string
+		)
+		err := row.Scan(&tagKey, &outputValue, &outputType, &returnType)
+		if err != nil {
+			t.Fatal(err)
+		}
+		if tagKey != "tag1" {
+			t.Errorf("Expected the tag keys to be equal: %v != %v", tagKey, "tag1")
+		}
+		if outputValue != tval.outputValue {
+			t.Errorf("Expected the output values to be equal: %v != %v", outputValue, tval.outputValue)
+		}
+		if outputType != tval.outputType {
+			t.Errorf("Expected the output types to be equal: %v != %v", outputType, tval.outputType)
+		}
+		if returnType != opr.returnType {
+			t.Errorf("Expected the return types to be equal: %v != %v", returnType, opr.returnType)
+		}
```

Comment:
you can use `require.Equal` to simplify this.

Reply:
Awesome!